### ISSUE TITLE
FanoutPublisherSink fails with unhelpful error message when default subscription timeout triggers

### ISSUE THREAD

**Issue by jrudolph** [akka/akka#28926]
As observed in https://github.com/akka/akka-http/issues/3039

The fanout including the timeout is also used in `preMaterialize`. 5 seconds after `preMaterialize` is called, the publisher silently shuts down and any later subscriber will get an obnoxious "Processor actor [Actor[akka://test-server/system/Materializers/StreamSupervisor-1/flow-0-1-fanoutPublisherSink#622867610]] terminated abruptly".

It took me a long while to figure that out because the 5 second default subscription timeout is longer than you would wait in tests (even interactive) for it to trigger.

It can be fixed by explicitly failing the future subscribers when the timeout triggers instead of relying on the too general `postStop` in ActorProcessorImpl.

**Comment by jrudolph** [comment]
Fix should go here:

https://github.com/akka/akka/blob/89f920a1983b0b718c8bbb519a5b518e76bab45b/akka-stream/src/main/scala/akka/stream/impl/FanoutProcessor.scala#L164

The question if the timeout should be disabled for `preMaterialize` in the first place or if it can be documented away.

**Comment by guntiso** [comment]
I'd vote for the fix of `preMaterialize`. The whole idea of **pre**Materialize seems to be to materialize **later**. Should not require extra `Sink.ignore` for that.

**Comment by alexdupre** [comment]
I encountered this issue as well. I think the `preMaterialize` should not create a fanout with a timeout.

**Comment by jrudolph** [comment]
> I encountered this issue as well. I think the `preMaterialize` should not create a fanout with a timeout.

The disadvantage would be that without a timeout you will have a memory leak if the prematerialized stream is never used.

**Comment by alexdupre** [comment]
Well, if you don't want to make it the default, at least there should be the possibility for the user to decide to have a timeout or not.

**Comment by johanandren** [comment]
I think it could be possible to disable the timeout already, using `StreamSubscriptionTimeoutTerminationMode.NoopTermination`, but there is no access to set attributes on that `Sink.asPublisher`/`Source.asSubscriber` since it is created and run inside the `preMaterialize()` method.

This also means that it may be impossible to set other attributes for the async boundary (buffer size for example) - maybe the stream should pass on attributes of its own to the subscriber/publisher operator?

**Comment by johanandren** [comment]
BTW not entirely sure perhaps this is somewhat related to #30908 where the timeout happens, and cancels a subscription, but what bubbles up is "you can't subscribe multiple times" for a late subscriber, rather than something saying "publisher already canceled".

**Comment by alexdupre** [comment]
I don't think that's related, in that case it was created with `fanout = false`, so there was no timeout as far as I understand.

**Comment by johanandren** [comment]
There are timeouts both for fanout and non-fanout publisher sinks, so the _somewhat_ related part is the failing with something that does not make it clear it is about the publisher having been cancelled.

**Comment by Christewart** [comment]
Encountering this as well

**Comment by johanandren** [comment]
`preMaterialize` is implemented only through public APIs so I think it would be possible to compose your own version with attributes configuring the timeout.

**Comment by nodefactory-bk** [comment]
Yep, that seems to have worked with some sligh copying of akka code =)
For the next person that might need it:

```  
def fancyPreMaterialize[Out,Mat](in: Source[Out,Mat], timeout: FiniteDuration)
                         (implicit materializer: Materializer)
      : (Mat, Source[Out, akka.NotUsed]) =
  {
    val timeoutAttr = ActorAttributes.streamSubscriptionTimeout(
      timeout,
      StreamSubscriptionTimeoutTerminationMode.cancel
    )

    val (mat, pub) = in
      .toMat(Sink.asPublisher(fanout = true))(Keep.both)
      .addAttributes(timeoutAttr)
      .run()

    mat -> Source.fromPublisher(pub)
  }
```

**Comment by Captain1653** [comment]
It looks like this issue can be closed. Did I get something wrong?

**Comment by johanandren** [comment]
Let's leave it around, there is a workaround but it is somewhat messy.

**Comment by johanandren** [comment]
That sounds about right, can happen in a highly loaded system where subscribing is delayed for some reason. Saturated CPU or starved thread pool because of blocking could explain it, but it could also be something else in a stream design delaying subscription.

**Comment by BillOTei** [comment]
> That sounds about right, can happen in a highly loaded system where subscribing is delayed for some reason. Saturated CPU or starved thread pool because of blocking could explain it, but it could also be something else in a stream design delaying subscription.

alright thanks, as we have mainly blocking requests on the API side, play's thread pool configuration is something we need to look into indeed.
